core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala

```diff
@@ -310,12 +343,41 @@ private[spark] class TaskSchedulerImpl(
       }
     }
 
-    // Randomly shuffle offers to avoid always placing tasks on the same set of workers.
-    val shuffledOffers = Random.shuffle(offers)
+    // ensure that we periodically check if executors can be removed from the blacklist, without
+    // requiring a separate thread and added synchronization overhead
```

Comment:
Can you change this comment to say something about doing this *before* making any offers? Maybe something like

Before making any offers, remove any nodes from the blacklist whose blacklist has expired.   Do this here to avoid a separate thread and added synchronization overhead, and also because updating the blacklist is only relevant when task offers are being made.